portinuse.c build error with OpenBSD 6.8

I uncommented /*#define CHECK_PORTINUSE*/ in the miniupnpd configure file and nothing else and attempted to build miniupnpd-2.2.0-RC2 from source code using OpenBSD 6.8. I used only the make command and the build process produced the following errors:

> cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o miniupnpd.o ./miniupnpd.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnphttp.o ./upnphttp.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpdescgen.o ./upnpdescgen.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpsoap.o ./upnpsoap.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpredirect.o ./upnpredirect.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o getifaddr.o ./getifaddr.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o daemonize.o ./daemonize.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpglobalvars.o ./upnpglobalvars.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o options.o ./options.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnppermissions.o ./upnppermissions.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o minissdp.o ./minissdp.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o natpmp.o ./natpmp.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o pcpserver.o ./pcpserver.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpevents.o ./upnpevents.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnputils.o ./upnputils.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o getconnstatus.o ./getconnstatus.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpstun.o ./upnpstun.c
./upnpstun.c:255:86: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
                syslog(LOG_DEBUG, "%s: waiting %ld secs and %ld usecs", "wait_for_stun_responses", timeout.tv_sec, timeout.tv_usec);
                                               ~~~                                                 ^~~~~~~~~~~~~~
                                               %lld
1 warning generated.
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnppinhole.o ./upnppinhole.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o asyncsendto.o ./asyncsendto.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o portinuse.o ./portinuse.c
./portinuse.c:166:9: warning: implicit declaration of function 'CIRCLEQ_FIRST' is invalid in C99 [-Wimplicit-function-declaration]
        next = CIRCLEQ_FIRST(&table.inpt_queue); /*TAILQ_FIRST(&table.inpt_queue);*/
               ^
./portinuse.c:166:7: warning: incompatible integer to pointer conversion assigning to 'struct inpcb *' from 'int' [-Wint-conversion]
        next = CIRCLEQ_FIRST(&table.inpt_queue); /*TAILQ_FIRST(&table.inpt_queue);*/
             ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./portinuse.c:174:10: warning: implicit declaration of function 'CIRCLEQ_NEXT' is invalid in C99 [-Wimplicit-function-declaration]
                next = CIRCLEQ_NEXT(&inpcb, inp_queue); /*TAILQ_NEXT(&inpcb, inp_queue);*/
                       ^
./portinuse.c:174:31: error: use of undeclared identifier 'inp_queue'
                next = CIRCLEQ_NEXT(&inpcb, inp_queue); /*TAILQ_NEXT(&inpcb, inp_queue);*/
                                            ^
3 warnings and 1 error generated.
*** Error 1 in /usr/ports/pobj/miniupnpd-2.2.0-RC2/miniupnpd-2.2.0-RC2 (Makefile:260 'portinuse.o')



Had to use quote insert quote instead of insert code because the code did not format properly for this build issue I'm submitting.

**Edit:** I'll get back to you on this. I removed kvm from the OpenBSD miniupnpd Makefile a couple of days ago and is probably the issue. Why checking if a port is in use for OpenBSD is dependent on kvm is beyond me. Seems to be a lack of coding logic for when kvm is not in use.

Since OpenBSD 5.5 release "All CIRCLEQ uses replaced with TAILQ". I'm not sure why I wasn't able to find that doing a google search earlier. Anyways, just for a test I replaced both CIRCLEQ_FIRST WITH TAILQ_FIRST and the portinuse.c did not produce and error although changing CIRCLEQ_NEXT TO TAILQ_NEXT produced an error. See my build below using miniupnpd master source code dated October 28th, 2020.

> cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o miniupnpd.o ./miniupnpd.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnphttp.o ./upnphttp.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpdescgen.o ./upnpdescgen.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpsoap.o ./upnpsoap.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpredirect.o ./upnpredirect.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o getifaddr.o ./getifaddr.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o daemonize.o ./daemonize.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpglobalvars.o ./upnpglobalvars.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o options.o ./options.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnppermissions.o ./upnppermissions.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o minissdp.o ./minissdp.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o natpmp.o ./natpmp.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o pcpserver.o ./pcpserver.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpevents.o ./upnpevents.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnputils.o ./upnputils.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o getconnstatus.o ./getconnstatus.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnpstun.o ./upnpstun.c
./upnpstun.c:255:86: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
                syslog(LOG_DEBUG, "%s: waiting %ld secs and %ld usecs", "wait_for_stun_responses", timeout.tv_sec, timeout.tv_usec);
                                               ~~~                                                 ^~~~~~~~~~~~~~
                                               %lld
1 warning generated.
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o upnppinhole.o ./upnppinhole.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o asyncsendto.o ./asyncsendto.c
cc  -I. -O2 -pipe  -Wall -W -Wstrict-prototypes -fno-common -c -o portinuse.o ./portinuse.c
./portinuse.c:175:8: error: called object type 'struct inpcb *' is not a function or function pointer
                next TAILQ_NEXT(&inpcb, inp_queue);     /*TAILQ_NEXT(&inpcb, inp_queue);*/
                ~~~~ ^
/usr/include/sys/queue.h:430:33: note: expanded from macro 'TAILQ_NEXT'
#define TAILQ_NEXT(elm, field)          ((elm)->field.tqe_next)
                                        ^
1 error generated.
*** Error 1 in /usr/ports/pobj/miniupnpd-5/miniupnpd-miniupnpd-5 (Makefile:260 'portinuse.o')


Thanks for looking into this!

--- a/miniupnpd/Changelog.txt
+++ b/miniupnpd/Changelog.txt
@@ -1,4 +1,7 @@
-$Id: Changelog.txt,v 1.467 2020/06/20 15:45:30 nanard Exp $
+$Id: Changelog.txt,v 1.469 2020/10/30 21:11:50 nanard Exp $
+
+2020/10/30:
+  OpenBSD: portinuse.c compatible with OpenBSD 5.5+
 
 2020/10/22:
   netfilter_nft: fix rule-cache update when using the same chain name in

--- a/miniupnpd/configure
+++ b/miniupnpd/configure
@@ -1,5 +1,5 @@
 #! /bin/sh
-# $Id: configure,v 1.111 2020/05/10 22:26:13 nanard Exp $
+# $Id: configure,v 1.113 2020/10/30 21:11:52 nanard Exp $
 # vim: tabstop=4 shiftwidth=4 noexpandtab
 #
 # miniupnp daemon
@@ -199,6 +199,10 @@ case $OS_NAME in
 		# onrdomain was introduced in OpenBSD 5.0
 		if [ $MAJORVER -ge 5 ]; then
 			echo "#define PFRULE_HAS_ONRDOMAIN" >> ${CONFIGFILE}
+		fi
+		# before OpenBSD 5.5 inpt_queue was CIRCLEQ
+		if [ $MAJORVER -lt 5 ] || [ $MAJORVER -eq 5 -a $MINORVER -lt 5 ]; then
+			echo "#define INPT_QUEUE_IS_CIRCLEQ" >> ${CONFIGFILE}
 		fi
 		FW=pf
 		echo "#define USE_IFACEWATCHER 1" >> ${CONFIGFILE}

--- a/miniupnpd/portinuse.c
+++ b/miniupnpd/portinuse.c
@@ -1,7 +1,7 @@
-/* $Id: portinuse.c,v 1.9 2019/09/24 11:49:28 nanard Exp $ */
+/* $Id: portinuse.c,v 1.11 2020/10/30 21:11:52 nanard Exp $ */
 /* vim: tabstop=4 shiftwidth=4 noexpandtab
  * MiniUPnP project
- * (c) 2007-2019 Thomas Bernard
+ * (c) 2007-2020 Thomas Bernard
  * http://miniupnp.free.fr/ or https://miniupnp.tuxfamily.org/
  * This software is subject to the conditions detailed
  * in the LICENCE file provided within the distribution */
@@ -163,7 +163,12 @@ static struct nlist list[] = {
 		kvm_close(kd);
 		return -1;
 	}
-	next = CIRCLEQ_FIRST(&table.inpt_queue); /*TAILQ_FIRST(&table.inpt_queue);*/
+	/* inpt_queue was CIRCLEQ_HEAD, it is TAILQ_HEAD since OpenBSD 5.5 */
+#ifdef INPT_QUEUE_IS_CIRCLEQ
+	next = CIRCLEQ_FIRST(&table.inpt_queue);
+#else
+	next = TAILQ_FIRST(&table.inpt_queue);
+#endif
 	while(next != NULL) {
 		if(((u_long)next & 3) != 0) break;
 		n = kvm_read(kd, (u_long)next, &inpcb, sizeof(inpcb));
@@ -171,7 +176,11 @@ static struct nlist list[] = {
 			syslog(LOG_ERR, "kvm_read(): %s", kvm_geterr(kd));
 			break;
 		}
-		next = CIRCLEQ_NEXT(&inpcb, inp_queue);	/*TAILQ_NEXT(&inpcb, inp_queue);*/
+#ifdef INPT_QUEUE_IS_CIRCLEQ
+		next = CIRCLEQ_NEXT(&inpcb, inp_queue);
+#else
+		next = TAILQ_NEXT(&inpcb, inp_queue);
+#endif
 		/* skip IPv6 sockets */
 		if((inpcb.inp_flags & INP_IPV6) != 0)
 			continue;

